<a href="https://colab.research.google.com/github/nileshrathod17/Online-Retail-Customer-Segmentation/blob/main/Online_Retail_Customer_Segmentation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Customer Segmentation for Online Retail store**
We endeavour to find the various customer segments using the online retail store's transaction data obtained from UCI Machine Learning Dataset repository. This is a transnational data set which contains all the transactions occurring between 01/12/2010 and 09/12/2011 for a UK-based and registered non-store online retail.

# **The need of customer segmentation:**
The differences in customers' behaviour, demographics, geographies, etc. help in classifying them in groups. Learning about different groups in the customer can help with following:

Target Marketing               

Client understanding

Optimal product placement

Searching for new customers

Revenue growth




# **Recency-Frequency-Monetary (RFM) model to determine customer value:**
The RFM model is quite useful model in retail customer segmentation where only the data of customer transaction is available. RFM stands for the three dimensions:

Recency – How recently did the customer purchase?

Frequency – How often do they purchase?

Monetary Value – How much do they spend?

A combination of these three attributes can be defined to assign a quantitative value to customers. e.g. A customer who recently bought high value products and transacts regularly is a high value customer.

# **Import Libraries and Data**

In [14]:
# Import libraries
import pandas as pd
import datetime as dt
import numpy as np
import re
# import libraries
import pandas  as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.preprocessing import LabelEncoder 
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.manifold import TSNE
from sklearn.cluster import KMeans

import collections
from collections import Counter
from collections import defaultdict
import itertools

from scipy import stats
from sklearn.metrics import silhouette_score

import matplotlib.pyplot as plt
import matplotlib.dates as mdates
import seaborn as sns
plt.style.use('seaborn')
%config InlineBackend.figure_format = 'retina'
%matplotlib inline

In [15]:
# Mount your drive and read the csv file. Make sure you are using the correct file path
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [27]:
file_path = '/content/drive/MyDrive/Colab Notebooks/Online Retail Customer Segmentation/Online Retail.xlsx'
data=pd.read_excel(file_path)

In [28]:
data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom

# **Attribute Information**

InvoiceNo: Invoice number. Nominal, a 6-digit integral number uniquely assigned to each transaction. If this code starts with letter 'c', it indicates a cancellation.

StockCode: Product (item) code. Nominal, a 5-digit integral number uniquely assigned to each distinct product.

Description: Product (item) name. Nominal.

Quantity: The quantities of each product (item) per transaction. Numeric.

InvoiceDate: Invoice Date and time. Numeric, the day and time when each transaction was generated.

UnitPrice: Unit price. Numeric, Product price per unit in sterling.

CustomerID: Customer number. Nominal, a 5-digit integral number uniquely assigned to each customer.

Country: Country name. Nominal, the name of the country where each customer resides.

In [20]:
data.nunique()

InvoiceNo      25900
StockCode       4070
Description     4223
Quantity         722
InvoiceDate    23260
UnitPrice       1630
CustomerID      4372
Country           38
dtype: int64

In [29]:
data.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

In [30]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[ns]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 33.1+ MB


# **Initial Findings**

**Potential data reversal.**

The max and min value of Quantity are both 80995; it could represent a reversal of data entry. However, the issue is that both the initial and reversal entry are retained in the dataset. Further investigation is needed to understand the nature and determine the best way to manage such data reversal.
**Potential indication of cancellation orders from negative UnitPrice.**

It is uncommon to have negative UnitPrice, as this would mean a cash outflow to a company. These transactions could represent cancelled orders by customers or bad-debt/write-off incurred by the business.

**Missing 25% of CustomerID.**

The missing unique identified of customers could post a problem as market/customer segmentation requires grouping each unique customer into a group.

# **Data Pre-Processing and Feature Engineering**

# **InvoiceDate**

**Separate Date and Time information from InvoiceDate**

The InvoiceDate column contains both date and time of the transaction. These data are separated into individual columns to facilitate future feature engineering and data manipulation.

In [31]:
# Split datetime from InvoiceDate
data['Date'] = data['InvoiceDate'].dt.date
#df['Date'] = pd.to_datetime(df['Date'], errors='coerce')
data['Time'] = data['InvoiceDate'].dt.time
# Remove InvoiceDate column
data.drop(['InvoiceDate'], axis=1, inplace=True)
data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

   UnitPrice  CustomerID         Country        Date      Time  
0       2.55     17850.0  United Kingdom  2010-12-01  08:26:00  
1       3.39     17850.0  United Kingdom  2010-12-01  08:26:00  
2       2.75     17850.0  United Kingdom  2010-12-01  08:26:00  
3       3.39     17850.0  United Kingdom  2010-12-01  08:26:00  
4       3.39     17850.0  United Kingdom  2010-12-01  08:26:00

# **InvoiceNo**
**Extract transaction status from InvoiceNo**

InvoiceNo contains both transaction status (i.e. having a 'C' denotes cancelled transaction) and transaction identifier (e.g. unique invoice number). This information could be extracted to facilitate further feature engineering.

In [34]:
# Separate order status and invoice number from InvoiceNo
data['CancelledOrder'] = data['InvoiceNo'].apply(
    lambda x: re.findall(r'[A-Z]', str(x))).apply(lambda x: pd.Series(x))
data['Invoice_No'] = data['InvoiceNo'].apply(
    lambda x: re.findall(r'\d+', str(x))).apply(lambda x: pd.Series(x))

# Remove old InvoiceNo column
data.drop(['InvoiceNo'], axis=1, inplace=True)

data.head()

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: DeprecationWarning: The default dtype for empty Series will be 'object' instead of 'float64' in a future version. Specify a dtype explicitly to silence this warning.
  This is separate from the ipykernel package so we can avoid doing imports until


StockCode                          Description  Quantity  UnitPrice  \
0    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6       2.55   
1     71053                  WHITE METAL LANTERN         6       3.39   
2    84406B       CREAM CUPID HEARTS COAT HANGER         8       2.75   
3    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6       3.39   
4    84029E       RED WOOLLY HOTTIE WHITE HEART.         6       3.39   

   CustomerID         Country        Date      Time CancelledOrder Invoice_No  
0     17850.0  United Kingdom  2010-12-01  08:26:00            NaN     536365  
1     17850.0  United Kingdom  2010-12-01  08:26:00            NaN     536365  
2     17850.0  United Kingdom  2010-12-01  08:26:00            NaN     536365  
3     17850.0  United Kingdom  2010-12-01  08:26:00            NaN     536365  
4     17850.0  United Kingdom  2010-12-01  08:26:00            NaN     536365

In [36]:
# Encode CancelledOrder
data['CancelledOrder'] = data['CancelledOrder'].astype('category')
data['CancelledOrder'].unique()

[NaN, 'C', 'A']
Categories (2, object): ['A', 'C']

In [37]:
def filter_row(data, column, criterion, operator='equal'):
    '''
    Filter rows based on specific condition
    '''
    if operator == 'equal':
        return data[data[column] == criterion]
    if operator == 'less':
        return data[data[column] <= criterion]
    if operator == 'more':
        return data[data[column] >= criterion]

def remove_row(data, column, criterion):
    '''
    Remove ros based on specific condition
    '''
    return data[data[column] != criterion]

In [38]:
# 'A' category is unexpected; printing out rows to investigate further
filter_row(data, 'CancelledOrder', 'A')

StockCode      Description  Quantity  UnitPrice  CustomerID  \
299982         B  Adjust bad debt         1   11062.06         NaN   
299983         B  Adjust bad debt         1  -11062.06         NaN   
299984         B  Adjust bad debt         1  -11062.06         NaN   

               Country        Date      Time CancelledOrder Invoice_No  
299982  United Kingdom  2011-08-12  14:50:00              A     563185  
299983  United Kingdom  2011-08-12  14:51:00              A     563186  
299984  United Kingdom  2011-08-12  14:52:00              A     563187

# **Drop Bad Debt Records**

Bad debt adjustments are dropped from the dataset as these do not represent actual sales. Furthermore, they are not tagged to any specific customer.

In [39]:
# Drop Bad Debt Record
data = remove_row(data, 'CancelledOrder', 'A')

# Encode column
data['CancelledOrder'] = data['CancelledOrder'].cat.add_categories([0])
data['CancelledOrder'].fillna(value=0, inplace=True)
data['CancelledOrder'].replace(to_replace='C', value=1, inplace=True)

data.head()

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """
/usr/local/lib/python3.7/dist-packages/pandas/core/generic.py:6392: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return self._update_inplace(result)
/usr/local/lib/python3.7/dist-packages/pandas/core/generic.py:6619: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  retu

StockCode                          Description  Quantity  UnitPrice  \
0    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6       2.55   
1     71053                  WHITE METAL LANTERN         6       3.39   
2    84406B       CREAM CUPID HEARTS COAT HANGER         8       2.75   
3    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6       3.39   
4    84029E       RED WOOLLY HOTTIE WHITE HEART.         6       3.39   

   CustomerID         Country        Date      Time CancelledOrder Invoice_No  
0     17850.0  United Kingdom  2010-12-01  08:26:00              0     536365  
1     17850.0  United Kingdom  2010-12-01  08:26:00              0     536365  
2     17850.0  United Kingdom  2010-12-01  08:26:00              0     536365  
3     17850.0  United Kingdom  2010-12-01  08:26:00              0     536365  
4     17850.0  United Kingdom  2010-12-01  08:26:00              0     536365